pkg/join/join.go

```diff
@@ -0,0 +1,216 @@
+/*
+SPDX-License-Identifier: Apache-2.0
+
+Copyright Contributors to the Submariner project.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package join
+
+import (
+	"fmt"
+	"github.com/submariner-io/submariner-operator/internal/cluster"
+	"github.com/submariner-io/submariner-operator/internal/constants"
+	"github.com/submariner-io/submariner-operator/internal/image"
+	"github.com/submariner-io/submariner-operator/internal/nodes"
+	"github.com/submariner-io/submariner-operator/internal/restconfig"
+	"github.com/submariner-io/submariner-operator/pkg/broker"
+	"github.com/submariner-io/submariner-operator/pkg/client"
+	"github.com/submariner-io/submariner-operator/pkg/deploy"
+	"github.com/submariner-io/submariner-operator/pkg/discovery/globalnet"
+	"github.com/submariner-io/submariner-operator/pkg/reporter"
+	"github.com/submariner-io/submariner-operator/pkg/secret"
+	"github.com/submariner-io/submariner-operator/pkg/version"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"strings"
+)
+
+func SubmarinerCluster(brokerInfo broker.Info, jo deploy.WithJoinOptions, restConfigProducer restconfig.Producer, status reporter.Interface, gatewayNode struct{Node string}) error {
```

Comment:
Since we join a broker...  Also pass a `client.Producer`.

```suggestion
func Broker(brokerInfo broker.Info, jo deploy.WithJoinOptions, clientProducer client.Producer, status reporter.Interface, gatewayNode struct{Node string}) error {
```

Reply:
> Since we join a broker...
> 
We join a submariner cluster to a broker cluster.
> Also pass a `client.Producer`.
> 
`restConfigProducer` is needed by `DetermineID` so passing it here and getting `clientProducer` at L66 below.
 
> Also I think we should define an `Options` struct separate from `deploy`.
Where?